src/gadget.rs

```diff
@@ -0,0 +1,175 @@
+//! Self-contained circuit implementations of various primitives.
+
+use std::fmt;
+
+use crate::{
+    arithmetic::FieldExt,
+    plonk::{Advice, Any, Column, ConstraintSystem, Error, Fixed},
+};
+
+pub mod sha256;
+
+/// Handles the configuration of a chip
+pub trait FloorPlanner {}
+
+/// The configuration for a chip that is relevant to a [`Layouter`].
+pub trait ChipConfig {
+    // lookup columns
+    // permutation columns
+}
+
+/// A chip implements a set of instructions that can be used by gadgets.
+///
+/// The chip itself should not store any state; instead, state that is required at circuit
+/// synthesis time should be stored in [`Chip::Config`], which can then be fetched via
+/// [`Layouter::config`].
+pub trait Chip: Sized {
+    /// A type that holds the configuration for this chip, and any other state it may need
+    /// during circuit synthesis.
+    type Config: ChipConfig;
+
+    /// The field that the chip is defined over.
+    ///
+    /// This provides a type that the chip's configuration can reference if necessary.
+    type Field: FieldExt;
+
+    /// Load any fixed configuration for this chip into the circuit.
+    fn load(layouter: &mut impl Layouter<Self>) -> Result<(), Error>;
+}
+
+/// A pointer to a cell within a circuit.
+#[derive(Clone, Copy, Debug)]
+pub struct Cell {
+    /// Identifies the region in which this cell resides.
+    region_index: usize,
+    row_offset: usize,
+    column: Column<Any>,
+}
+
+/// A permutation configured by a chip.
+#[derive(Debug)]
+pub struct Permutation {
+    index: usize,
+    mapping: Vec<Column<Any>>,
+}
+
+impl Permutation {
+    /// Configures a new permutation for the given columns.
+    pub fn new<F: FieldExt>(meta: &mut ConstraintSystem<F>, columns: &[Column<Advice>]) -> Self {
+        let index = meta.permutation(columns);
+        Permutation {
+            index,
+            mapping: columns.iter().map(|c| (*c).into()).collect(),
+        }
+    }
+}
+
+/// This trait allows a [`Chip`] to direct a [`Layouter`] to assign cells within a
+/// region.
+///
+/// TODO: It would be great if we could constrain the columns in these types to be
+/// "logical" columns that are guaranteed to correspond to the chip (and have come from
+/// `Chip::Config`).
```

Comment:
One way to do this would be to have `DynRegion` and the chip parameterized on the type of columns that it uses. Then it would only be able to refer to its own columns in places where that type variable is used (even though it is always instantiated as Column<Any>, it's still treated as a different type per chip instance).

Unfortunately this is is complicated by the categorization of columns. It's still possible, for example we could use `DynRegion<CHIP: Chip, CELL: Cell, CFixed: Column<Fixed>, CAdvice: Column<Advice>>`. (This variant also constrains Cells to be in the right column.)

Reply:
Another possibility, that is probably cleaner now that I think of it, is to parameterize on a Region `R`, i.e. use `Cell<R>` and `Column<R, {Any, Fixed, Advice, Aux}>`.

Reply:
A chip might contains multiple regions, but if this only needs to be a distinguisher then we could use the chip itself as the "region" type here (or a concrete type that it knows).

Reply:
We'd also need to be able to refer to columns that are *not* constrained to a chip. I think maybe that pushes the complexity of the `Cell<R>` and `Column<R, {Any, Fixed, Advice, Aux}>` approach above that of the `DynRegion<CHIP: Chip, CELL: Cell, CFixed: Column<Fixed>, CAdvice: Column<Advice>>` approach.